src/main/java/org/opensearch/neuralsearch/query/HybridQuery.java

```diff
@@ -106,22 +111,40 @@ public Query rewrite(IndexSearcher indexSearcher) throws IOException {
             return new MatchNoDocsQuery("empty HybridQuery");
         }
 
-        boolean actuallyRewritten = false;
-        List<Query> rewrittenSubQueries = new ArrayList<>();
+        final HybridQueryRewriteCollectorManager manager = new HybridQueryRewriteCollectorManager(indexSearcher);
+        final List<Callable<Void>> queryRewriteTasks = new ArrayList<>();
+        final List<HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>>> collectors = new ArrayList<>();
         for (Query subQuery : subQueries) {
-            Query rewrittenSub = subQuery.rewrite(indexSearcher);
-            /* we keep rewrite sub-query unless it's not equal to itself, it may take multiple levels of recursive calls
-               queries need to be rewritten from high-level clauses into lower-level clauses because low-level clauses
-               perform better. For hybrid query we need to track progress of re-write for all sub-queries */
-            actuallyRewritten |= rewrittenSub != subQuery;
-            rewrittenSubQueries.add(rewrittenSub);
+            final HybridQueryExecutorCollector<IndexSearcher, Map.Entry<Query, Boolean>> collector = manager.newCollector();
```

Comment:
Just for my understanding, why is a collector chosen here? you could have plugged in TaskExecutor here and then simply use the list of results returned by invokeAll. The entire collector seems to be a round about way to get the results. Manager would have been useful if you wanted to return a different collector based on certain conditions IMO

Reply:
Good question. The main idea is to move logic of collecting and reducing/merging into its own manager. By introducing TaskExecutor inside Scorer/Weight we are coupling how to parallelize with what needs to be parallelized. With introduction of Collector and Collector Manager, we decoupled parallelization from logic. IMO, this abstracts parallelization implementation from all of its caller.

Reply:
Lets dissect this a little more. There are 3 classes here:
1. HybridQueryExecutor: responsible for initializing and holding a common threadpool for query, scorer and weight,
2. HybridQueryExecutorCollectorManager: Responsible for giving a new collector
3. HybridQueryExecutorCollector : Holds the query results

> With introduction of Collector and Collector Manager, we decoupled parallelization from logic. IMO, this abstracts parallelization implementation from all of its caller.

HybridQueryExecutor seems to be a good wrapper to make sure the same task executor is used. I am more curious about the additional value of `HybridQueryExecutorCollectorManager` and `HybridQueryExecutorCollector`

Why can't it be
```
final List<Callable<Entry<Query, Boolean>>> queryRewriteTasks = new ArrayList<>();
// for each subquery
queryRewriteTasks.add(() -> rewriteQuery(subQuery); //rewrite needs to return entry<query, boolean>;
List<Entry<Query, Boolean>> rewrittenQueries = HybridQueryExecutor.getExecutor().invokeAll(queryRewriteTasks);

// rest of the logic
```
This way you don't have to worry about the threadsafety of the collectors. There are some utility methods in collector manager which arent related to managing a collector but more of related to rewritten queries itself.

Overall while this works fine, I want to make sure these interfaces have a clear definition and its clear on how to use them. Currently it seems like we can simplify this by not having it unless I am misunderstanding the value it brings